source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include "source/extensions/config_subscription/grpc/grpc_stream.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * This class arbitrates between two config providers of the same GrpcMux -
+ * the primary and the failover. Envoy always prefers fetching config from the
+ * primary source, but if not available, will fetch the config from the failover
+ * source until the primary is again available.
+ *
+ * This class owns the state for the GrpcMux primary and failover streams, and
+ * proxies the gRPC stream functionality to either the primary or the failover config sources.
+ * The failover source is optional and will only be used if passed upon initialization.
+ *
+ * The primary config source is always preferred over the failover source. If the primary
+ * is available, the stream to the failover source will terminated.
+ *
+ * Failover is supported in both
+ * SotW (RequestType = envoy::service::discovery::v3::DiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DiscoveryResponse), and
+ * Delta-xDS (RequestType = envoy::service::discovery::v3::DeltaDiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DeltaDiscoveryResponse).
+ * Both the primary and failover streams are either SotW or Delta-xDS.
+ *
+ * The use of this class will be as follows: the GrpcMux object will own an instance of
+ * the GrpcMuxFailoverManager. The GrpcMuxFailoverManager will own 2 GrpcStreams, primary
```

Comment:
question: we talk about `GrpcMuxFailoverManager` here but the class below is named `GrpcMuxFailover`, and the description implies this is the class that owns the 2 GrpcStreams.  Should this be rewritten with GrpcMuxFailover instead of GrpcMuxFailoverManager?

Reply:
Good catch... trying to figure out the naming properly across many drafts. I think GrpcMuxFailover is the better one. Updated.

Reply:
This is a better name but still not updated here :P